### PR TITLE
[DPE-4683] Add URI

### PR DESCRIPTION
**Pull request by dragomirp**
Adds URI to relation data

### CODE REVIEW

File: src/relations/pgbouncer_provider.py [hunk 1]
```diff
@@ -228,24 +229,43 @@ def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
...
+
+        host = self.charm.leader_ip if exposed else "localhost"
+        port = self.charm.config["listen_port"]
+        uri_host = quote(
+            ",".join([
+                self.charm.leader_ip,
+                *[ip for ip in self.charm.peers.units_ips if ip != self.charm.leader_ip],
+            ])
+            if exposed
+            else f"/tmp/snap-private-tmp/snap.charmed-pgbouncer/tmp/{self.charm.app.name}/instance_0",
```
Comment (dragomirp) [inline src/relations/pgbouncer_provider.py:252]: Socket won't work with snapped clients. Plugs TBD.

File: tests/integration/relations/pgbouncer_provider/helpers.py [hunk 2]
```diff
@@ -176,15 +176,19 @@ def check_exposed_connection(credentials, tls):
...
-    connstr = f"dbname='{database}' user='{user}' host='{host}' port='{port}' password='{password}' connect_timeout=1 sslmode={sslmode}"
+    if "uris" in credentials["postgresql"]:
+        uri = credentials["postgresql"]["uris"]
+        connstr = f"{uri}?connect_timeout=1&sslmode={sslmode}"
+    else:
+        host, port = credentials["postgresql"]["endpoints"].split(":")
+        user = credentials["postgresql"]["username"]
+        password = credentials["postgresql"]["password"]
+        database = credentials["postgresql"]["database"]
+        connstr = f"dbname='{database}' user='{user}' host='{host}' port='{port}' password='{password}' connect_timeout=1 sslmode={sslmode}"
```
Comment (dragomirp) [inline tests/integration/relations/pgbouncer_provider/helpers.py:191]: Data integrator upgrade test still relies on manually constructing the connection string

File: src/relations/pgbouncer_provider.py [hunk 3]
```diff
@@ -228,28 +229,48 @@ def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
...
+        )
+        rw_endpoint = f"{host}:{port}"
         self.database_provides.set_endpoints(
             relation.id,
             rw_endpoint,
         )
+        # Set connection string URI.
+        self.database_provides.set_uris(
+            relation.id,
+            f"postgresql://{user}:{password}@{quote(uri_host, safe=',')}:{port}/{database}",
```
Comment (dragomirp) [inline src/relations/pgbouncer_provider.py:272]: Not escaping the `,` for better readability.

File: src/relations/pgbouncer_provider.py [hunk 4]
```diff
@@ -228,28 +229,48 @@ def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
...
+
+        host = "localhost"
+        uri_host = (
+            f"/tmp/snap-private-tmp/snap.charmed-pgbouncer/tmp/{self.charm.app.name}/instance_0"
         )
         if exposed:
             if vip := self.charm.config.get("vip"):
-                ip = vip
+                host = vip
+                uri_host = vip
```
Comment (dragomirp) [inline src/relations/pgbouncer_provider.py:251]: If the vip is set only use the vip

File: src/relations/pgbouncer_provider.py [hunk 5]
```diff
@@ -228,28 +229,48 @@ def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
...
             else:
-                ip = self.charm.leader_ip
-            rw_endpoint = f"{ip}:{self.charm.config['listen_port']}"
-        else:
-            rw_endpoint = f"localhost:{self.charm.config['listen_port']}"
+                host = self.charm.leader_ip
+                uri_host = ",".join([
+                    self.charm.leader_ip,
+                    *[ip for ip in self.charm.peers.units_ips if ip != self.charm.leader_ip],
+                ])
```
Comment (dragomirp) [inline src/relations/pgbouncer_provider.py:257]: If exposed without a vip, collect all the ips in the uri.

**Review by taurus-forever — APPROVED** [review]
Really nice! Tnx!

**Review by marceloneppel — APPROVED** [review]

**Pull request merged**